examples/mixture/plot_gmm_selection.py

```diff
@@ -3,64 +3,85 @@
-but BIC is better suited if the problem is to identify the right model.
-Unlike Bayesian procedures, such inferences are prior-free.
+This example shows that model selection can be performed with Gaussian Mixture
+Models using :ref:`information-theory criteria <aic_bic>`. Model selection
+concerns both the covariance type and the number of components in the model.
 
-In that case, the model with 2 components and full covariance
-(which corresponds to the true generative model) is selected.
+In this case, both the Akaike Information Criterion (AIC) and the Bayes
+Information Criterion (BIC) provide the right result, but we only demo the
+latter as BIC is better suited to identify the true model among a set of
+candidates. Unlike Bayesian procedures, such inferences are prior-free.
 
 """
 
-import numpy as np
-import itertools
-
-from scipy import linalg
-import matplotlib.pyplot as plt
-import matplotlib as mpl
+# %%
+# Data generation
+# ---------------
+#
+# We generate two components (each one containing `n_samples`) by randomly
+# sampling the standard normal distribution as returned by `numpy.random.randn`.
+# One component is kept spherical yet shifted and re-scaled. The other one is
+# deformed to have a more general covariance matrix.
 
-from sklearn import mixture
+import numpy as np
 
-# Number of samples per component
 n_samples = 500
-
-# Generate random sample, two components
 np.random.seed(0)
 C = np.array([[0.0, -0.1], [1.7, 0.4]])
+
 X = np.r_[
     np.dot(np.random.randn(n_samples, 2), C),
     0.7 * np.random.randn(n_samples, 2) + np.array([-6, 3]),
 ]
 
+# %%
+# Model training and selection
+# ----------------------------
+#
+# We vary the number of components from 1 to 6 and the type of covariance
+# parameters to use:
+#
+# - `"full"`: each component has its own general covariance matrix.
+# - `"tied"`: all components share the same general covariance matrix.
+# - `"diag"`: each component has its own diagonal covariance matrix.
+# - `"spherical"`: each component has its own single variance.
+#
+# We score the different models and keep the best model (the lowest BIC).
+
+from sklearn.mixture import GaussianMixture
+
 lowest_bic = np.infty
 bic = []
 n_components_range = range(1, 7)
-cv_types = ["spherical", "tied", "diag", "full"]
-for cv_type in cv_types:
+cov_types = ["spherical", "tied", "diag", "full"]
+for cov_type in cov_types:
     for n_components in n_components_range:
         # Fit a Gaussian mixture with EM
-        gmm = mixture.GaussianMixture(
-            n_components=n_components, covariance_type=cv_type
-        )
+        gmm = GaussianMixture(n_components=n_components, covariance_type=cov_type)
         gmm.fit(X)
         bic.append(gmm.bic(X))
         if bic[-1] < lowest_bic:
             lowest_bic = bic[-1]
             best_gmm = gmm
 
 bic = np.array(bic)
-color_iter = itertools.cycle(["navy", "turquoise", "cornflowerblue", "darkorange"])
 clf = best_gmm
```

Comment:
I think that we can improve the example using a proper `GridSearchCV` here:

```python
from sklearn.model_selection import GridSearchCV


def gmm_bic_score(estimator, X):
    """Callable to pass to GridSearchCV that will use the BIC score."""
    # Make it negative since GridSearchCV expects a score to maximize
    return -estimator.bic(X)


param_grid = {
    "n_components": range(1, 7),
    "covariance_type": ["spherical", "tied", "diag", "full"],
}
grid_search = GridSearchCV(GaussianMixture(), param_grid=param_grid, scoring=gmm_bic_score)
grid_search.fit(X)
```

It shows how to define a callable for the grid-search.
The nice part is that `grid_search` will default to the `best_estimator_` after `fit` is called.